enhancements/20220615-seeder.md

```diff
@@ -0,0 +1,267 @@
+  * `Inventory`: Inventory is a wrapper around the `BaseboardManagement` CRD, along with some additional information needed by the `boots` and the Harvester installer. Inventory will be a long lived object on the cluster, and will be associated with an actual metal node. `Inventory` will be associated with a `Cluster` object in seeder, and will be assigned to the cluster for the lifetime of the cluster object. On cluster deletion the Inventory object will be freed and returned back to the pool for reuse. The `Inventory` object manages an associated `BaseboardManagement` object, which can be used to perform tasks on the underlying metal hardware. The inventory controller performs the following core tasks:
+    * On creation of `Inventory` object, create an associated `BaseboardManagement` object and then reconciles the state based on underlying bmc connectivity and power state reported by `rufio`. 
+    * When `Inventory` object is allocated to a cluster, based on specific conditions, such as `Hardware` object generation condition, the controller creates a `BMCJob` object, which is used by `rufio` to trigger reboot of the associated metal node. This eventually triggers DHCP/PXE boot serve to kick start Harvester installation.
+    * When `Inventory` object is freed from a cluster, the controller triggers another `BMCJob` to power off the metal node.
+  * `Cluster`: Cluster is an abstraction of an actual Harvester cluster running on a series of baremetal nodes managed by associated `Inventory` objects. The cluster controller performs the following core tasks:
+    * On creation of a Cluster, allocate a `Harvester` VIP from the associated AddressPool.
+    * Using a combination of `Inventory` CRD spec and `Harvester` CRD spec, generate a `Hardware` CRD, with a valid userdata. This userdata is passed as kernel arguments by `boots` as part of the PXE booting workflow, and can be used to perform all Harvester configuration via kernel arguments.
+    * Patch the `Inventory` status with correct conditions, allowing `Inventory` controller to trigger the reboot of the nodes via `BMCJobs` as describe above. 
+    * Reconcile `Inventory` objects in `Cluster` spec, allowing users to resize the cluster by add / remove inventory as needed.
+
+### User Stories
+Once Seeder is deployed to a k8s cluster, the users should be able to quickly provision Harvester clusters.
+
+#### Story 1
+Create a Harvester Cluster
+
+By submitting a yaml file, a user will be able to have a full functioning multi-node harvester cluster available for use.
+
+Sample yaml file:
+```yaml
+apiVersion: metal.harvesterhci.io/v1alpha1
+kind: AddressPool
+metadata:
+  name: node-pool
+  namespace: default
+spec:
+  cidr: "172.16.128.11/32"
+  gateway: "172.16.128.1"
+  netmask: "255.255.248.0"
```

Comment:
This is weird
CIDR has IP and netmask
and we have CIDR and netmask at the same time here?

Reply:
In the actual CRD spec, netmask is optional. 

For example in this case I want to allocate a pre-existing IP address but want to override the netmask calculated from cidr range. 

https://github.com/harvester/seeder/blob/main/pkg/api/v1alpha1/addresspool_types.go#L34-L38

Reply:
It might confuse others.
because CIDR and netmask describe the same attribute of IP.
So in this example, what will we see in Linux `ip addr` command?